include/lo2s/perf/reader.hpp

```diff
@@ -0,0 +1,368 @@
+     */
+    PerfEventInstance open(std::variant<Cpu, Thread> location, int group_fd = -1,
+                           int cgroup_fd = config().cgroup_fd);
+
+    perf_event_attr& get_attr()
+    {
+        return attr_;
+    }
+
+    double get_scale() const
+    {
+        return scale_;
+    }
+
+    void set_scale(double scale)
+    {
+        scale_ = scale;
+    }
+
+    std::string get_unit() const
+    {
+        return unit_;
+    }
+
+    void set_unit(std::string unit)
+    {
+        unit_ = unit;
+    }
+
+    friend bool operator==(const PerfEvent& lhs, const PerfEvent& rhs)
+    {
+        return !memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+    friend bool operator<(const PerfEvent& lhs, const PerfEvent& rhs)
+    {
+        return memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+    ~PerfEvent()
+    {
+    }
+
+protected:
+    struct perf_event_attr attr_;
+    double scale_ = 1;
+    std::string unit_ = "#";
+};
+
+/**
+ * Contains predefined events, like the ones listed by --list-events
+ * @note define the type by either calling as_counter() or as_group_leader() after creating an
+ * object
+ */
+class PredefEvent : public PerfEvent
+{
+public:
+    PredefEvent();
+    PredefEvent(bool enable_on_exec, const std::string& ev_name);
+    PredefEvent(std::string name, perf_type_id type, std::uint64_t config,
+                std::uint64_t config1 = 0, std::set<Cpu> cpus = std::set<Cpu>());
+
+    const std::set<Cpu>& supported_cpus() const;
+
+    // automatically determine availability
+    void set_availability();
+
+    const Availability& get_availability() const
+    {
+        return availability_;
+    };
+
+    bool is_valid() const;
+    bool is_available_in(ExecutionScope scope) const
+    {
+        // per-process should always work. the counter will just not count if the process is
+        // scheduled on a core that is not supprted by that counter
+        return scope.is_thread() || cpus_.empty() || cpus_.count(scope.as_cpu());
+    }
+
+    std::string get_name() const
+    {
+        return name_;
+    }
+
+    std::set<Cpu> get_cpus() const
+    {
+        return cpus_;
+    }
+
+    void set_common_attrs(std::string name, perf_type_id type, std::uint64_t config,
+                          std::uint64_t config1 = 0, std::set<Cpu> cpus = std::set<Cpu>())
+    {
+        name_ = name;
+        cpus_ = cpus;
+        attr_.type = type;
+        attr_.config = config;
+        attr_.config1 = config1;
+    }
+
+    void as_counter();
+    void as_group_leader();
+
+private:
+    std::string name_ = "";
+    std::set<Cpu> cpus_;
+    Availability availability_ = Availability::UNAVAILABLE;
+};
+
+/**
+ * Contains an event parsed from sysfs
+ * @note call on as_sample() after creation to get a valid
+ * event, otherwise the availability will be set to UNAVAILABLE
+ */
+class SysfsEvent : public PredefEvent
```

Comment:
Move everything in sysfs_read_event (src/perf/event_provider) into the constructor of SysfsEvent